src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -0,0 +1,246 @@
+        new(@"^[a-zA-Z0-9\.\-_]{1,255}$", RegexOptions.Compiled, TimeSpan.FromMilliseconds(100));
+
+    public KafkaTransportPump(KafkaTransportBindingElement transportBindingElement,
+        ILogger<KafkaTransportPump> logger,
+        IServiceDispatcher serviceDispatcher, KafkaDeliverySemantics kafkaDeliverySemantics)
+    {
+        _logger = logger;
+        _kafkaDeliverySemantics = kafkaDeliverySemantics;
+        Topic = serviceDispatcher.BaseAddress.PathAndQuery.TrimStart('/');
+        if (string.IsNullOrEmpty(Topic) || !s_topicNameRegex.IsMatch(Topic))
+        {
+            throw new NotSupportedException(string.Format(SR.InvalidTopicName, Topic));
+        }
+
+        if (transportBindingElement.ErrorHandlingStrategy == KafkaErrorHandlingStrategy.DeadLetterQueue)
+        {
+            if (string.IsNullOrEmpty(transportBindingElement.DeadLetterQueueTopic))
+            {
+                throw new NotSupportedException(SR.InvalidDeadLetterQueueTopicName);
+            }
+
+            if (!s_topicNameRegex.IsMatch(transportBindingElement.DeadLetterQueueTopic))
+            {
+                throw new NotSupportedException(string.Format(SR.InvalidTopicName, Topic));
+            }
+        }
+        TransportBindingElement = transportBindingElement;
+        _baseAddress = serviceDispatcher.BaseAddress;
+    }
+
+    public override Task StartPumpAsync(QueueTransportContext queueTransportContext, CancellationToken token)
+    {
+        _cts = new();
+        _mres = new();
+        _mres.Reset();
+        _receiveContextCountdownEvent = new(1);
+
+        CancellationToken ct = CancellationTokenSource
+            .CreateLinkedTokenSource(_cts.Token, token).Token;
+
+        ConsumerConfig = TransportBindingElement.Config;
+        ConsumerConfig.BootstrapServers = _baseAddress.Authority;
+        var (enableAutoCommit, enableAutoOffsetStore) = GetCommitStrategyConfigValues(ConsumerConfig, _kafkaDeliverySemantics);
+        ConsumerConfig.EnableAutoCommit = enableAutoCommit;
+        ConsumerConfig.EnableAutoOffsetStore = enableAutoOffsetStore;
+        Consumer = new ConsumerBuilder<Null, byte[]>(ConsumerConfig)
+            .SetKeyDeserializer(Deserializers.Null)
+            .SetValueDeserializer(Deserializers.ByteArray)
+            .SetLogHandler(OnLog)
+            .SetErrorHandler(OnError)
+            .Build();
+
+        Consumer.Subscribe(Topic);
+
+        if (TransportBindingElement.ErrorHandlingStrategy == KafkaErrorHandlingStrategy.DeadLetterQueue)
+        {
+            ProducerConfig producerConfig = new(ConsumerConfig);
+            producerConfig.Acks = Acks.All;
+            Producer = new ProducerBuilder<Null, byte[]>(producerConfig)
+                .SetKeySerializer(Serializers.Null)
+                .SetValueSerializer(Serializers.ByteArray)
+                .Build();
+        }
+
+        Task.Run(async () =>
+        {
+            while (!ct.IsCancellationRequested)
+            {
```

Comment:
This only has a concurrency of 1. This can become a bottleneck if the messages can be large. Is this intentional? A limitation of the Kafka consumer client?

Reply:
Yes. Under the hood multiple threads are concurrently retrieving the messages. Please see https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#the-consume-loop

Reply:
thank you for this link, understood the background processing to fetch messages... but once messages are fetched aren't we consuming/processing one by one ? (via await OnConsumeMessage). I don't have the expertise on how kafka messages are processed around service, but if this a normal trend, I am fine. Let me know what you think.

Reply:
Indeed we are processing the message one by one. With Kafka all consumers from a given consumer group are given offset of a topic. Each consumer is committing its offset (i.e something like ack in rabbitmq) so messages have to be consumed sequentially by each consumer. If one need to scale out because the consumer loop is too slow then that's need to add a consumer to the consumer group. In the dotnet sdk the exposed abstraction is consumer that's why the consume loop is always single threaded. (i've never seen a kafka dotnet consumer sample multithreaded to be honest). When scaling out the consumer group by adding a new consumer to the group please notice that the topic need to have at least one partition to distribute by consumer.